code/game/machinery/doors/firedoor.dm

```diff
@@ -23,12 +23,86 @@
 	interaction_flags_machine = INTERACT_MACHINE_WIRES_IF_OPEN | INTERACT_MACHINE_ALLOW_SILICON | INTERACT_MACHINE_OPEN_SILICON | INTERACT_MACHINE_REQUIRES_SILICON | INTERACT_MACHINE_OPEN
 	var/nextstate = null
 	var/boltslocked = TRUE
-	var/list/affecting_areas
 	var/being_held_open = FALSE
+	/// The merger id used to create/get the merger group in charge of handling tanks that share an internal gas storage
+	var/merger_id = "firelock"
+	///Don't register the turfs signal more than once
+	var/registered_turfs = FALSE
+	///Stops the firealarm from reopening the door if there is an hazard nearby
+	var/hazard = FALSE
 
 /obj/machinery/door/firedoor/Initialize()
 	. = ..()
-	CalculateAffectingAreas()
+	var/area/my_area = get_area(src)
+	LAZYADD(my_area.firedoors, src)
+
+	RegisterSignal(src, COMSIG_MERGER_ADDING, .proc/MergerAdding)
+	RegisterSignal(src, COMSIG_MERGER_REMOVING, .proc/MergerRemoving)
+
+	return INITIALIZE_HINT_LATELOAD
+
+/obj/machinery/door/firedoor/LateInitialize()
+	. = ..()
+	GetMergeGroup(merger_id, allowed_types = typecacheof(/obj/machinery/door/firedoor))
+
+/obj/machinery/door/firedoor/Destroy()
+	var/area/my_area = get_area(src)
+	LAZYREMOVE(my_area.firedoors,src)
+	return..()
+
+///////////////////////////////////////////////////////////////////
+// Merger handling
+
+/obj/machinery/door/firedoor/proc/MergerAdding(obj/machinery/door/firedoor/us, datum/merger/new_merger)
+	SIGNAL_HANDLER
+	if(new_merger.id != merger_id)
+		return
+	RegisterSignal(new_merger, COMSIG_MERGER_REFRESH_COMPLETE, .proc/MergerRefreshComplete)
+
+/obj/machinery/door/firedoor/proc/MergerRemoving(obj/machinery/door/firedoor/us, datum/merger/old_merger)
+	SIGNAL_HANDLER
+	if(old_merger.id != merger_id)
+		return
+	UnregisterSignal(old_merger, COMSIG_MERGER_REFRESH_COMPLETE)
+
+/// Handles the firelocks to register only one signal per group
+/obj/machinery/door/firedoor/proc/MergerRefreshComplete(datum/merger/merger, list/leaving_members, list/joining_members)
+	SIGNAL_HANDLER
+	var/turf/open/turf1 = get_step(src, dir)
+	var/turf/open/turf2 = get_step(src, turn(dir, 180))
+	if(merger.origin != src)
+		registered_turfs = FALSE
+		UnregisterSignal(turf1, COMSIG_TURF_EXPOSE)
+		UnregisterSignal(turf2, COMSIG_TURF_EXPOSE)
+		return
+
+	if(!registered_turfs)
+		RegisterSignal(turf1, COMSIG_TURF_EXPOSE, .proc/check_closing)
+		RegisterSignal(turf2, COMSIG_TURF_EXPOSE, .proc/check_closing)
```

Comment:
Trusting dir like this is a bad idea I think. Hard to think of a better way of handling it though.
Listening to adjacent open tiles seems sane. I think that'll work in most places, though it might break with space adjacent firelocks? you'd need to check.

Reply:
maybe i'll add a space check and prevent the signal to be registered on space (there shouldn't be walls in turf1 and turf2 if mapper do their job correctly, but maybe i'll add a warning for it)

Reply:
I don't think mappers angle firelocks. just like doors in general, if it doesn't already have a visual or behavioral reason to be angled, it won't be. Plus a firelock should activate if the turf to its side gets cold, not just the front and back. However you define front any back

Reply:
the turfs on the side of the firelocks are either walls, dense or have another firelock that does the same, you just need to check the two turfs, not four.

Reply:
also in the next commit i'll add the openturf check